Support building a mono-based .NET Runtime on x64

omajid opened the pull request:
We now have some architectures (eg, s390x and ppc64le) that produce a .NET runtime that looks and feels like any other .NET runtime, except it's using Mono instead of CoreCLR under the hood.

However, it's only possible to produce this Mono-based .NET runtime on s390x/ppc64le. That makes it hard to test this set up on other architectures. Issues that affect the mono build on all architectures - such as #66594 become harder to fix and verify because of this unnecessary architecture requirement.

Fix that by adding a flag to the top-level build.sh (and also to the msbuild projects) to make it possible to produce a .NET runtime with Mono on any platform.

The default configuration is unchanged: we still produced CoreCLR-based .NET runtime on x64 and a Mono-based runtime on s390x/ppc64le.

Fixes: #62440

omajid commented:
cc @directhex

tmds commented on eng/Subsets.props:
```diff
@@ -26,8 +26,8 @@
   <PropertyGroup>
-    <PrimaryRuntimeFlavor>CoreCLR</PrimaryRuntimeFlavor>
-    <PrimaryRuntimeFlavor Condition="'$(TargetArchitecture)' == 's390x' or '$(TargetArchitecture)' == 'armv6'">Mono</PrimaryRuntimeFlavor>
+    <PrimaryRuntimeFlavor Condition="'$(PrimaryRuntimeFlavor)' == '' and ('$(TargetArchitecture)' == 's390x' or '$(TargetArchitecture)' == 'armv6')">Mono</PrimaryRuntimeFlavor>
+    <PrimaryRuntimeFlavor Condition="'$(PrimaryRuntimeFlavor)' == ''">CoreCLR</PrimaryRuntimeFlavor>
```
You can simplify the condition by splitting it across a few lines.
```xml
<DefaultPrimaryRuntimeFlavor>CoreCLR<DefaultPrimaryRuntimeFlavor/>
<DefaultPrimaryRuntimeFlavor Condition="'$(TargetArchitecture)' == 's390x'">Mono</DefaultPrimaryRuntimeFlavor>
<DefaultPrimaryRuntimeFlavor Condition="'$(TargetArchitecture)' == 'armv6'">Mono</DefaultPrimaryRuntimeFlavor>
<PrimaryRuntimeFlavor Condition="'$(PrimaryRuntimeFlavor)' == ''">$(DefaultPrimaryRuntimeFlavor)</PrimaryRuntimeFlavor>

```

ViktorHofer reviewed (commented):
Can you please document the PrimaryRuntimeFlavor switch? Especially, I would be interested in a section that explains when to use PrimaryRuntimeFlavor vs RuntimeFlavor.

omajid commented:
Thanks for the feeback!

> Can you please document the PrimaryRuntimeFlavor switch?

FWIW, this PR doesn't add it; the property already existed. We do make it possible for users to override it. The `eng/Subsets.props` documents it like this already:

```
  <!-- Determine the primary runtime flavor. This is usually CoreCLR, except on
       platforms (like s390x) where only Mono is supported. The primary runtime
       flavor is used to decide when to build the hosts and installers. -->
```

And the `--help` also adds:

```
  echo "  -primaryRuntime                 Primary Runtime: CoreCLR or Mono. Unlike -runtimeFlavor, this produces a .NET-like"
  echo "                                  runtime (including tools, apphosts, etc) but uses either Mono or CoreCLR as the"
  echo "                                  actual runtime. Useful for building .NET on architectures without CoreCLR support,"
  echo "                                  or emulating the build for those architectures on x64."
```

What sort of detail are you looking for? 

What would be a good place to document it?

omajid commented:
Thanks to @tmds' comment (https://github.com/dotnet/runtime/issues/62440#issuecomment-1108138345), I have learned that we don't need to adjust `PrimaryRuntimeFlavor` in the props file, we can simply pass in the correct value via msbuild property flag through build.sh. Is there any reason to add the conditionals to the .props file? I guess it makes things more explicit?

ViktorHofer commented on eng/build.sh:
```diff
@@ -38,6 +38,10 @@ usage()
   echo "                                  [Default: Debug]"
   echo "  -runtimeFlavor (-rf)            Runtime flavor: CoreCLR or Mono."
   echo "                                  [Default: CoreCLR]"
+  echo "  -primaryRuntime                 Primary Runtime: CoreCLR or Mono. Unlike -runtimeFlavor, this produces a .NET-like"
```
Don't we need a corresponding change in build.ps1 for Windows?

omajid replied:
Yes, we do. I need to learn Windows/Powershell :'(

omajid replied:
I don't have access to a Windows environment to fully test my changes, but I have used the powershell docker container image and `Set-PSDebug -Trace 1` to confirm that 

    ./eng/build.ps1 -primaryRuntime Mono -v d

Results in:

    DEBUG:    1+  >>>> & "/runtime/eng/common/build.ps1" /p:TargetArchitecture=x64 -restore -build /p:PrimaryRuntimeFlavor=mono -verbosity d -configuration Debug

omajid committed:
Support building a mono-based .NET Runtime on x64

We now have some architectures (eg, s390x) that produce a .NET runtime
that looks and feels like any other .NET runtime, except it's using
Mono instead of CoreCLR under the hood.

However, it's only possible to produce this Mono-based .NET runtime on a
select few architectures. That makes it hard to test this set up on
other architectures.  Issues that affect the mono build on all
architectures - such as #66594 become harder to fix and verify because
of this unnecessary architecture requirement.

Fix that by adding a flag to the top-level build.sh (and also to the
msbuild projects) to make it possible to produce a .NET runtime with
Mono on any platform.

The default configuration is unchanged: we still produced CoreCLR-based
.NET runtime on x64 and a Mono-based runtime on arm32,ppc64le,s390x.

Fixes: #62440

ViktorHofer commented:
There's a merge conflict in the PR. Apart from that, looks good.

tmds commented on eng/build.sh:
```diff
@@ -38,6 +38,10 @@ usage()
+  echo "  -primaryRuntime                 Primary Runtime: CoreCLR or Mono. Unlike -runtimeFlavor, this produces a .NET-like"
+  echo "                                  runtime (including tools, apphosts, etc) but uses either Mono or CoreCLR as the"
+  echo "                                  actual runtime. Useful for building .NET on architectures without CoreCLR support,"
+  echo "                                  or emulating the build for those architectures on x64."
```
fwiw, I still don't like this name. I had rather had an argument that captures _this produces a .NET runtime_.

We  now have:

To build a .NET-like runtime that uses CoreCLR on x64:
```
./build.sh
```
To build mono on x64:
```
./build.sh -runtimeFlavor Mono
```
To build a .NET-like runtime that uses mono on x64:
```
./build.sh -runtimeFlavor Mono -primaryRuntime Mono
```
Undefined behavior:
```
./build.sh -primaryRuntime Mono
```

akoeplinger replied:
I agree. Do we really need a separate argument or could we say if you explicitly specify a runtimeFlavor then make that the primary runtime?

akoeplinger replied:
looks like this PR stalled a bit, what's our current thinking here?

omajid replied:
I am not sure what the way forward here is.

As far as a I can tell, `./build.sh -primaryRuntime Mono` is the same as `./build.sh -runtimeFlavor Mono`: you get mono and nothing else.

I see @tmds's point about making things more complex, but I am not sure what the desired way forward is. As far as I can

I think making runtimeFlavor the primary runtime is the opposite of what we want. We want to be able to produce a .NET runtime (what this PR was calling primaryRuntime) where the underlying runtime (aka runtimeFlavor) is mono, not CoreCLR.

tmds replied:
Suggestion: add a flag named `--mono-runtime` with description `Use the Mono runtime instead of the CoreCLR runtime.`.

The build scrip can use this to set `PrimaryRuntimeFlavor`/`RuntimeFlavor` as needed.

For the build script user, I think provides a user-friendly option.

tmds replied:
@akoeplinger what do you think about the suggestion?

omajid replied:
I like the idea of `--mono-runtime`! I think "runtime" is a little ambiguous, though - I think we are building the .NET runtime with the Mono VM (instead of the Mono runtime with the Mono VM).

akoeplinger replied:
Sounds good. I think we could shorten it to just `--mono`.

tmds replied:
Let's use the same name as source-build: `--use-mono-runtime`.

https://github.com/dotnet/dotnet/blob/2fc4462a30e08fe5eccb6dbc271a4297d2bba75f/build.sh#L12

steveisok commented:
@omajid @tmds @akoeplinger This has been open for a while.  Outside of resolving conflicts, what else needs resolved?

omajid commented:
> Outside of resolving conflicts, what else needs resolved?

I need to update this PR to use the new argument name that we have converged on.

akoeplinger commented:
@omajid let me know if you need help resolving the conflicts :)

lewing commented:
This is the oldest open pr now :)

tmds commented on eng/build.sh:
```diff
@@ -360,6 +361,11 @@ while [[ $# > 0 ]]; do
       shift 2
       ;;
 
+     -usemonoruntime)
```
The vmr option is [--use-mono-runtime](https://github.com/dotnet/dotnet/blob/ea11d078482522dfa30c5caae1a890a14f6ac909/build.sh#L12). Can you use the exact same name?

omajid replied:
Good point! The rest of the options in the runtime build.sh use a completely different convention, though (eg `outputrid`, `keepnativesymbols`). Whatever I go with, it will look out of place with the other convention :(

Since this option lives in build.sh, I tried to keep it consistent with the other options in build.sh.

I am happy to change it if it's okay with the runtime maintainers.

lewing replied:
Sounds reasonable to me

tmds replied:
I'm fine leaving out the dashes. I didn't realize you had done it to be consistent with the other arguments.

akoeplinger replied:
I'm fine with the current state too.

marek-safar commented:
@directhex @akoeplinger could you please help landing this PR

akoeplinger commented:
Thank you!
